common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
         forever: false
       }
     }, amqp10.Policy.EventHub));
+
+    this._fsm = new machina.Fsm({
+      namespace: sdkVersionString,
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (disconnectCallback, err, result) => {
+            if (err) {
+              this._safeCallback(disconnectCallback, err);
```

Comment:
>this._safeCallback(disconnectCallback, err); [](start = 14, length = 44)

(This is more of a meta question.  My examples in here my be not correct, but I suspect we'll see this pattern all over the code and I want to make sure it's right)
I may be wrong here, but the parameter "disconnectCallback" doesn't necessarily come from a call to disconnect(), right?  If the connection gets disconnected calling some other function, the code _might_ end up here with disconnectCallback being the callback that was passed into that other function.  If that's the case, should we pass results.Disconnected() as the third parameter to the callback so that the caller is able to tell the final result of that operation.  Or are they responsible for figuring out that their operation caused a disconnect via some other means (such as checking the connection status)?

Reply:
they shouldn't care that the connection is disconnected. Also, events signaling the connection state will come in a later feature. so they will have a separate way to know. using results.Disconnected here kindof breaks the pattern of having either an error or a result but not both.

---
In reply to: [139503297](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139503297) [](ancestors = 139503297)